influxdb/metric.go

```diff
@@ -0,0 +1,229 @@
+	endTime := timeNow().Add(-m.offsetDuration)
+
+	resp, err := c.Query(m.buildQuery(startTime, endTime))
+	if err != nil {
+		return nil, nil, err
+	} else if err = resp.Error(); err != nil {
+		return nil, nil, err
+	}
+
+	if len(resp.Results) != 1 {
+		return nil, nil, fmt.Errorf("InfluxDB query '%s' returned %d query results, expected 1", m.config.Query, len(resp.Results))
+	}
+
+	if len(resp.Results[0].Series) == 0 {
+		log.Infof(ctx, "InfluxDB query '%s' returned no time series", m.config.Query)
+		return nil, nil, nil
+	} else if len(resp.Results[0].Series) > 1 {
+		return nil, nil, fmt.Errorf("InfluxDB query '%s' returned %d time series", m.config.Query, len(resp.Results[0].Series))
+	}
+
+	serie := resp.Results[0].Series[0]
+	points, err := parseSeriePoints(serie)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to parse InfluxDB points: %v", err)
+	}
+
+	// For gauge metrics, we don't need to apply any filtering since we only
+	// query InfluxDB for new points, and points that are too fresh are ignored
+	// by applying an offset. We'll need to apply filtering once cumulative
+	// metrics come into play.
+	timeSeries, err := m.convertTimeSeries(points)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to convert InfluxDB points %v to time series: %v", points, err)
+	}
+
+	return m.metricDescriptor(), timeSeries, nil
+}
+
+// buildQuery creates a InfluxDB query from the metric query definition,
+// wrapped in the given time-interval, inclusive of start time.
+func (m *Metric) buildQuery(startTime, endTime time.Time) client.Query {
+	query := fmt.Sprintf(
+		"SELECT * FROM (%s) WHERE time >= %d AND time < %d",
+		m.config.Query,
+		startTime.UnixNano(),
+		endTime.UnixNano())
+
+	return client.NewQuery(query, m.config.Database, "ns")
+}
+
+func (m *Metric) metricDescriptor() *metricpb.MetricDescriptor {
+	return &metricpb.MetricDescriptor{
+		Type:        m.StackdriverName(),
+		MetricKind:  metricpb.MetricDescriptor_GAUGE,
+		ValueType:   metricpb.MetricDescriptor_DOUBLE,
+		Description: fmt.Sprintf("InfluxDB query: %s", m.Name),
+		DisplayName: m.config.Query,
+	}
+}
+
+func (m *Metric) convertTimeSeries(points []point) ([]*monitoringpb.TimeSeries, error) {
+	ts := make([]*monitoringpb.TimeSeries, 0, len(points))
+	for _, p := range points {
+		newP, err := p.convertPoint()
+		if err != nil {
+			return nil, err
+		}
+
+		ts = append(ts, &monitoringpb.TimeSeries{
+			Metric:     &metricpb.Metric{Type: m.StackdriverName()},
+			Resource:   &monitoredres.MonitoredResource{Type: "global"},
+			MetricKind: metricpb.MetricDescriptor_GAUGE,
+			ValueType:  metricpb.MetricDescriptor_DOUBLE,
+			Points:     []*monitoringpb.Point{newP},
+		})
+	}
+	return ts, nil
+}
+
+type point struct {
+	timestamp time.Time
+	value     float64
+}
+
+// parseSeriePoints parses points from an InfluxDB series into a slice of
+// timestamp-value pairs.
+func parseSeriePoints(serie models.Row) ([]point, error) {
```

Comment:
Is this domain-specific lingo? Why are all these named "serie" instead of "series"?

Reply:
O I thought the singular form of "series" was "serie". Updated.